File: Sorting Algorithms/insertion_sort.py
import random


def main():
    test_array = []
    for i in range(100):
        test_array.append(random.randint(0, 100))
    sorted_array = insertion_sort(test_array.copy())
    r_sorted_array = reverse_insertion_sort(test_array.copy())

    print(f'Original List:         {test_array}')
    print(f'Sorted List:           {sorted_array}')
    print(f'Reverse Sorted List:   {r_sorted_array}')
    # Checksums
    print(f'Original List Checksum:         {sum(test_array)}')
    print(f'Sorted List Checksum:           {sum(sorted_array)}')
    print(f'Reverse Sorted List Checksum:   {sum(r_sorted_array)}')


def insertion_sort(array):
    if len(array) > 1:
        for i in range(len(array) - 1):
            key = array[i + 1]  # index of the current key (so i is the prior key)
            while i >= 0 and array[i] > key:  # get the index of prior keys until prior key is smaller
                array[i + 1] = array[i]
                i -= 1
            array[i + 1] = key  # keep key in position if larger than all prior keys
        return array
    else:
        print(f'Warning: array is too short to sort!')
        return array


def reverse_insertion_sort(array):
    if len(array) > 1:
        for i in range(len(array) - 1):
            key = array[i + 1]  # index of the current key (so i is the prior key)
            while i >= 0 and array[i] < key:  # get the index of prior keys until prior key is larger
                array[i + 1] = array[i]
                i -= 1
            array[i + 1] = key  # keep key in position if smaller than all prior keys
        return array
    else:
        print(f'Warning: array is too short to sort!')
        return array


if __name__ == '__main__':
    main()

File: Sorting Algorithms/Insertion Sort/insertion_sort.cpp

#include <iostream>
#include <cstdlib>


const int maxValue = 100;
const int testArraySize = 25;


void InsertionSort(int array[], int arraySize){
    int key;
    for (int i = 0; i < arraySize - 1; i++) {
        key = array[i + 1];  // value of index of the current key, i is the key before
        while (i >= 0 && array[i] > key)  // so long as i is valid index, loop until prior index's value is smaller
        {
            array[i + 1] = array[i];
            i--;
        }
        array[i + 1] = key;
    }
}


void ReverseInsertionSort(int array[], int arraySize){
    int key;
    for (int i = 0; i < arraySize - 1; i++) {
        key = array[i + 1];  // value of index of the current key, i is the key before
        while (i >= 0 && array[i] < key)  // so long as i is valid index, loop until prior index's value is greater
        {
            array[i + 1] = array[i];
            i--;
        }
        array[i + 1] = key;
    }
}


int Checksum(int array[], int arraySize){
    int checksum = 0;
    for (int i = 0; i < arraySize; i++) {
        checksum += array[i];
    }
    return checksum;
}


int main(){

    int testArray[testArraySize];    
    int arraySize = sizeof(testArray) / sizeof(int);
    for (int i = 0; i < arraySize; i++) {
        // srand(51 * 7 * i * i);  // Used to try different random seeds
        testArray[i] = rand() % maxValue;
    }
    
    int sortedArray[testArraySize];
    //DuplicateArray(testArray, sortedArray, arraySize);
    std::copy(std::begin(testArray), std::end(testArray), std::begin(sortedArray));
    InsertionSort(sortedArray, arraySize);

    int reversedArray[testArraySize];
    std::copy(std::begin(testArray), std::end(testArray), std::begin(reversedArray));
    ReverseInsertionSort(reversedArray, arraySize);


    std::cout << "Original List:         " << "[";
    for (int i = 0; i < arraySize - 1; i++) {
        std::cout << testArray[i] << ", ";
    }
    std::cout << testArray[arraySize - 1] << "]" << std::endl;

    std::cout << "Sorted List:           " << "[";
    for (int i = 0; i < arraySize - 1; i++) {
        std::cout << sortedArray[i] << ", ";
    }
    std::cout << sortedArray[arraySize - 1] << "]" << std::endl;

    std::cout << "Reverse Sorted List:   " << "[";
    for (int i = 0; i < arraySize - 1; i++) {
        std::cout << reversedArray[i] << ", ";
    }
    std::cout << reversedArray[arraySize - 1] << "]" << std::endl;


    std::cout << "Original List Checksum:         " << Checksum(testArray, arraySize) << std::endl;
    std::cout << "Sorted List Checksum:           " << Checksum(sortedArray, arraySize) << std::endl;
    std::cout << "Reverse Sorted List Checksum:   " << Checksum(reversedArray, arraySize) << std::endl;

}
